Update ElasticSearch and  Elastica to 6.x

anvit opened the pull request:
Update ElasticSearch and Elastica and address issues introduced by breaking changes in ElasticSearch 6.x.

The most notable changes are:
- [Dropped support for multiple mapping types](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/breaking-changes-6.0.html#_indices_created_before_6_0): ES 6.x no longer allows using multiple indices, They suggest handling this by creating multiple indices, each with its own type.
- [Mapping changes](https://www.elastic.co/guide/en/elasticsearch/reference/6.8/breaking-changes-6.0.html#breaking_60_mappings_changes): The `include_in_all` mapping parameter is disallowed and the `_all` field is disabled by default. 

Changes in Elastica:
- Elastica only has a addDocuments, deleteDocuments, and updateDocuments methods in 6.x and there are no longer addDocument, deleteDocument, updateDocument methods for handling a singular documents.

anvit committed:
Update ElasticSearch and Elastica dependencies

Update ElasticSearch to 6.8.23 in the Dockerfile and update Elastica to
6.x in composer.

melaniekung committed:
Remove arElasticSearchIndexDecorator.

- 'inline' scripts are deprecated, changed to 'source'

anvit committed:
Update index for ES 6.x

Update arElasticSearchPlugin to use multiple indices instead of multiple
types since ES 6.x removed being able to add multiple types. Also update
mapping to remove include_in_all and add that to the _all field using
copy_to instead since include_in_all was also removed in ES 6.x

anvit committed:
Fix autocompleteAction for ES 6.x

Update autocompleteAction to use multiple indices

anvit committed:
Add index types for ES Queries without a type

Add the appropriate index types for ES requests that did not need to
specify explicit types in ES 5.x

anvit committed:
Fix updateByQuery calls for ElasticSearch 6.x

anvit committed:
Fix broken publication status update

Add missing document type to partialUpdate in arElasticSearchPlugin

anvit committed:
Fix linting errors

anvit committed:
Switch elasticsearch to the oss docker image

jraddaoui commented on apps/qubit/modules/search/actions/autocompleteAction.class.php:
```diff
@@ -76,7 +76,10 @@ public function execute($request)
             $search = new \Elastica\Search($client);
-            $search->addIndex($index)->addType($index->getType($item['type']));
+            foreach ($indices as $type => $index) {
+                $itemType = QubitSearch::getInstance()->index->getIndexTypeName($item['type']);
+                $search->addIndex($index)->addType($index->getType($itemType));
```
Instead of using a different type on each index, I'd suggest to use the dummy type name `_doc`, as it will be used in 7.x:

https://www.elastic.co/blog/removal-of-mapping-types-elasticsearch#schedule

Using the same type for all indices should simplify many places modified in this PR.

jraddaoui commented on lib/QubitLftSyncer.class.php:
```diff
@@ -92,8 +92,9 @@ private function repairEsChildrenLftValues()
-        $bulk->setType('QubitInformationObject');
+        $type = 'QubitInformationObject';
+        $bulk->setIndex(QubitSearch::getInstance()->index->getType($type)->getName());
+        $bulk->setType(QubitSearch::getInstance()->index->getIndexTypeName($type));
```
This would be just `$bulk->setType("_doc");`

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchIndexDecorator.class.php:
```diff
@@ -17,67 +17,66 @@
+    public function __construct($prefix)
     {
-        $this->_instance = $instance;
+        $this->_instance = [];
```
It feels strange to use an index decorator to manage multiple indexes. I'd remove this class and move the logic altogether to `arElasticSearchPlugin.class.php`.

anvit replied:
I looked at removing this at first but realized that one reason we'd kept this file was that since we introduced multiple indices, we needed to use the ES index name set in the AtoM config as the prefix for all index names, and we would be repeating a lot of code for adding in the prefix if we didn't include a function for doing this.

I've renamed this class and refactored some code to clarify what this class is for.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPluginUtil.class.php:
```diff
@@ -482,8 +482,8 @@ protected static function getAllObjectStringFields(
                     (
-                        !isset($propertyProperties['include_in_all'])
-                        || $propertyProperties['include_in_all']
+                        !isset($propertyProperties['copy_to'])
+                        || ('_all' == $propertyProperties['copy_to'])
```
This condition will now copy to `_all` all the mapping fields where you removed `include_in_all: false`. It may be better to add `copy_to: _all` to the fields we really want to include and change this condition to `isset($propertyProperties['copy_to']) && ('_all' == $propertyProperties['copy_to'])`.

anvit replied:
Updated the mappings.yml file to copy all relevant mappings to _all

jraddaoui reviewed (changes requested):
Looking great @anvit!

I have not been able to look at everything in detail yet, but I have added a few comments that I think are important and may require some work already. Let me know your thoughts ;)

anvit committed:
Update mapping condition for _all fields

Update condition in arElasticSearchPluginUtil to only get string
fields included in _all if copy_to is explicitly set to _all in the
mappping.

anvit committed:
Update ES mapping to copy relevant fields to _all

anvit committed:
Rename arElasticSearchIndexDecorator

Renamed arElasticSearchIndexDecorator as
arElasticSearchMultiIndexWrapper for clarity and added a few comments
about changes needed for ElasticSearch/Elastica 7.x.

anvit committed:
Rename getType function from ESMultiIndexWrapper

Rename getType from arElasticSearchMultiIndexWrapper to getIndex for
clarity.

anvit committed:
Add dummy type for ES index

Add a dummy ElasticSearch index type since one is still required for
ES 6.x, instead of having a seperate type name for each of the
multiple indices.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -44,6 +52,13 @@ class arElasticSearchPlugin extends QubitSearchEngine
      */
     public $index;
```
It's not longer an Elastic_Index and it could be a good opportunity to name it `indicesWrapper` or something like that, since you are already changing many of the places where it's used.

anvit replied:
We're planning on retaining the old name for backwards compatibility of custom themes for now. Will probably update this in the future when we refactor away ElasticSearch's use in AtoM that's outside the Search/ElasticSearch plugin.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -44,6 +52,13 @@ class arElasticSearchPlugin extends QubitSearchEngine
+     *
+     * @var mixed defaults to null
+     */
+    protected $currentBatchType;
```
Maybe `$currentBatchIndex`, `$currentBatchIndexName` or `$currentBatchName` is better?

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -90,9 +105,7 @@ public function __construct(array $options = [])
-        // TODO: arElasticSearchIndexDecorator can be removed when Elastica >= 6.x. It is present to
-        // provide Elastica/Index::updateByQuery().
-        $this->index = new arElasticSearchIndexDecorator($this->client->getIndex($this->config['index']['name']));
+        $this->index = new arElasticSearchMultiIndexWrapper($this->config['index']['name']);
```
I think you don't need the prefix in the wrapper ...

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
+        foreach ($this->mappings as $typeName => $typeProperties) {
+            $typeName = 'Qubit'.sfInflector::camelize($typeName);
+            $this->index->createIndex($typeName,
+                $this->client->getIndex($this->index->getIndexName($typeName))
```
... it may only be needed here. Removing it from the wrapper will simplify things a bit, just using the default index names internally.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
-                );
-            }
+                // Load and normalize mappings
+                $this->loadAndNormalizeMappings();
```
Why calling this again inside the loop?

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,89 @@
+ */
+class arElasticSearchMultiIndexWrapper
+{
+    protected $_instance;
```
I'd call this `indices`. Also, why the underscore?

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,89 @@
+{
+    protected $_instance;
+
+    protected $_indexPrefix;
```
As I said, you could leave the prefix outside of the wrapper. Use the default names internally and append the prefix only when calling ES. Removing the need of `getIndexName`.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,89 @@
+        $this->_indexPrefix = $prefix;
+    }
+
+    public function createIndex($name, Elastica\Index $index)
```
I'd call this function `addIndex`.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,89 @@
+        return $this->_instance[$name];
+    }
+
+    public function getInstance()
```
I'd call it `getIndices`.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,89 @@
+    {
+        return $this->_instance;
+    }
+}
```
Considering all already said, I think you could live without this class and move the remaining pieces to arElasticSearchPlugin. Having `indices` there directly instead of the wrapper.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -30,6 +30,14 @@ class arElasticSearchPlugin extends QubitSearchEngine
      */
     public const MIN_VERSION = '1.3.0';
```
Should this be updated?

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -343,10 +356,23 @@ public function addDocument($data, $type)
+        }
 
         if ($this->batchMode) {
             // Add this document to the batch add queue
```
This comment is misplaced now.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -343,10 +356,23 @@ public function addDocument($data, $type)
+            if ($this->currentBatchType != $type) {
+                $this->flushBatch();
+                $this->currentBatchType = $type;
+                $this->index->refresh();
```
It would be nice to be able to pass an index name to flush only that one, now that we are flushing when we change the index. Defaults to refresh them all.

If you remove the wrapper and have access to the `indices` here, you could even refresh it directly (`$this->indices[$name]->refresh();`) and only do the full refresh over all in the `__destruct` function.

jraddaoui replied:
Sorry, I meant refresh when I said flush. This is my suggestion to avoid refreshing all indices when we only flush the batch in one of them:

* Looking at the calls of `$this->index->refresh();` they all happen after `$this->flushBatch();`. So I'd move that call inside the flushBatch method and remove it from all other places.
* Now that you are sure it only happens there, you also know that we only flush the batch for one index, so you could change the call to include the index name: `$this->index->refresh($this->currentBatchIndexName);`.
* And the refresh function in the wrapper could require a `$name` parameter and only refresh that index:
  
  ```php
  public function refresh($name)
  {
      $this->indices[$name]->refresh();
  }
  ```

Additionally, similar to the comment I made about the update methods, up to you if you want to include it in the wrapper or not, you could do it outside like you are doing now for the updates: `$this->index->getIndex($this->currentBatchIndexName)->refresh();`. I think it's important to be consistent so, now that we are going with the wrapper, I think it makes more sense to put everything there.

jraddaoui replied:
I think what I suggested above:

> Looking at the calls of $this->index->refresh(); they all happen after $this->flushBatch();. So I'd move that call inside the flushBatch method and remove it from all other places.

It would have simplified the latest changes and the code in general, only one call to refresh inside flushBatch, and always using `$this->currentBatchIndexName`, no need to make the index name optional since it always happens per index.

Sorry if that wasn't clear, and feel free to leave it as it is ;)

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -427,12 +459,24 @@ public function delete($object)
+            if ($this->currentBatchType != $type) {
+                $this->flushBatch();
+                $this->currentBatchType = $type;
+                $this->index->refresh();
```
Same here.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
+
+                    if ($indexType != $this->index->getIndexName($typeName)) {
+                        continue;
+                    }
```
This foreach and if seem unnecessary, you should be able to access the normalized mapping related to the index by using the index name directly `$this->mappings[$indexName]`.

jraddaoui replied:
I'll follow up about this in another comment.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
-                $mapping->setProperties($typeProperties['properties']);
+                    // Setting a dummy type since it is required in ES 6.x
+                    // but it can be removed in 7.x when it becomes optional
+                    $mapping->setType($index->getType(self::ES_TYPE));
```
It should be enough with `$mapping->setType(self::ES_TYPE);`.

anvit replied:
`mapping->setType` expects an `Elastica\Type`, which is why we're using `getType`:


https://github.com/ruflin/Elastica/blob/f95cf646f884a8e7b73bd12eda3e5753994b24d3/lib/Elastica/Index.php#L70-L80

jraddaoui replied:
I see, thanks!

jraddaoui commented on apps/qubit/modules/search/actions/autocompleteAction.class.php:
```diff
@@ -76,7 +76,12 @@ public function execute($request)
+                $itemType = QubitSearch::getInstance()::ES_TYPE;
+
+                // This can be updated in ES 7.x when type params are optional
+                $search->addIndex($index)->addType($index->getType($itemType));
```
Same, we should be able to use the constant directly, right? `$search->addIndex($index)->addType(QubitSearch::getInstance()::ES_TYPE);`

jraddaoui replied:
Addressed in another comment.

jraddaoui commented on lib/QubitLftSyncer.class.php:
```diff
@@ -92,8 +92,9 @@ private function repairEsChildrenLftValues()
-        $bulk->setType('QubitInformationObject');
+        $type = 'QubitInformationObject';
+        $bulk->setIndex(QubitSearch::getInstance()->index->getIndex($type)->getName());
+        $bulk->setType(QubitSearch::getInstance()->index->getIndexName($type));
```
Shouldn't this be constant? `$bulk->setType(QubitSearch::getInstance()::ES_TYPE);`

jraddaoui commented on lib/QubitLftSyncer.class.php:
```diff
@@ -92,8 +92,9 @@ private function repairEsChildrenLftValues()
-        $bulk->setIndex(QubitSearch::getInstance()->index->getName());
-        $bulk->setType('QubitInformationObject');
+        $type = 'QubitInformationObject';
+        $bulk->setIndex(QubitSearch::getInstance()->index->getIndex($type)->getName());
```
We may need the prefix here too. Right now, it's getting the index name from ES directly but you could do something like `$bulk->setIndex(QubitSearch::getInstance()->prefix + 'QubitInformationObject');`, if you add the prefix to arElasticSearchPlugin and make it public.

jraddaoui replied:
Addressed in a following comment.

jraddaoui commented on apps/qubit/modules/search/actions/autocompleteAction.class.php:
```diff
@@ -76,7 +76,12 @@ public function execute($request)
         foreach ($items as $item) {
             $search = new \Elastica\Search($client);
-            $search->addIndex($index)->addType($index->getType($item['type']));
+            foreach ($indices as $type => $index) {
```
I'd try to avoid the use of `$type`, I'd use $name in this case.

jraddaoui commented on lib/QubitLftSyncer.class.php:
```diff
@@ -92,8 +92,9 @@ private function repairEsChildrenLftValues()
         $bulk = new Elastica\Bulk(QubitSearch::getInstance()->client);
-        $bulk->setIndex(QubitSearch::getInstance()->index->getName());
-        $bulk->setType('QubitInformationObject');
+        $type = 'QubitInformationObject';
```
No need for a variable, if you still want one I'd call it `$name`.

jraddaoui commented on lib/job/arUpdatePublicationStatusJob.class.php:
```diff
@@ -91,7 +91,8 @@ public function runJob($parameters)
         ];
 
-        $response = QubitSearch::getInstance()->index->updateByQuery($query, $queryScript, $options)->getData();
+        $type = 'QubitInformationObject';
```
Same, no need for a var, if you want it I think it's better called `$name`.

jraddaoui commented on lib/task/tools/updatePublicationStatusTask.class.php:
```diff
@@ -166,14 +166,15 @@ protected function updatePublicationStatusDescendants($resource, $publicationSta
         $options = ['conflicts' => 'proceed'];
 
-        $response = QubitSearch::getInstance()->index->updateByQuery($query, $queryScript, $options)->getData();
+        $type = 'QubitInformationObject';
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -343,10 +356,23 @@ public function addDocument($data, $type)
+        $document->setType(self::ES_TYPE);
+
+        if (!$this->currentBatchType) {
+            $this->currentBatchType = $type;
```
We may even want to change the param name to `$indexName` instead of `$type`.

jraddaoui replied:
Also, unrelated to these changes, shouldn't this function be checking if it's enabled? Like all the others updates and delete.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -379,10 +405,16 @@ public function partialUpdate($object, $data)
             return;
         }
 
+        $type = get_class($object);
```
Call it `$name` or `$indexName`.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -427,12 +459,24 @@ public function delete($object)
         }
 
         if ($this->batchMode) {
+            $type = get_class($object);
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
-                        }
+        // Iterate over types (actor, informationobject, ...)
+        foreach ($this->mappings as $typeName => $typeProperties) {
+            $typeName = 'Qubit'.sfInflector::camelize($typeName);
```
`$indexName`

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
+        }
 
-                        $this->config['index']['configuration']['analysis']['analyzer'][$key]['char_filter'] = $filters;
+        foreach ($this->index->getInstance() as $indexType => $index) {
```
`$indexName`

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +537,86 @@ protected function initialize()
-            // Load and normalize mappings
-            $this->loadAndNormalizeMappings();
+                // Iterate over types (actor, informationobject, ...)
+                foreach ($this->mappings as $typeName => $typeProperties) {
```
`$indexName`, `$indexProperties`

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -61,22 +61,22 @@ mapping:
-      end_date_string: { type: keyword }
-      type_id: { type: integer, include_in_all: false }
-      actor_id: { type: integer, include_in_all: false }
+      start_date: { type: date, copy_to: _all }
+      end_date: { type: date, copy_to: _all }
```
I think these two are wrong and they should not have `copy_to: _all`.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -91,8 +91,8 @@ mapping:
     properties:
-      date: { type: date, include_in_all: false }
-      date_string: { type: keyword }
+      date: { type: date, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -102,111 +102,111 @@ mapping:
-      rights_holder: { type: text, include_in_all: false }
-      rights_note: { type: text, include_in_all: false }
-      license_terms: { type: text, include_in_all: false }
+      basis: { type: text, copy_to: _all }
+      start_date: { type: date, copy_to: _all }
+      end_date: { type: date, copy_to: _all }
+      copyright_status: { type: text, copy_to: _all }
+      rights_holder: { type: text, copy_to: _all }
+      rights_note: { type: text, copy_to: _all }
+      license_terms: { type: text, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -102,111 +102,111 @@ mapping:
-      restriction: { type: text, include_in_all: false }
-      start_date: { type: date, include_in_all: false }
-      end_date: { type: date, include_in_all: false }
+      act: { type: text, copy_to: _all }
+      restriction: { type: text, copy_to: _all }
+      start_date: { type: date, copy_to: _all }
+      end_date: { type: date, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -102,111 +102,111 @@ mapping:
+      codec_extensions_usually_used: { type: keyword, copy_to: _all }
+      file_size: { type: long, copy_to: _all }
+      duration: { type: integer, copy_to: _all }
+      bit_rate: { type: integer, copy_to: _all }
+      bit_rate_mode: { type: keyword, copy_to: _all }
+      overall_bit_rate: { type: integer, copy_to: _all }
+      channels: { type: integer, copy_to: _all }
+      channel_positions: { type: keyword, copy_to: _all }
+      sampling_rate: { type: integer, copy_to: _all }
+      samples_count: { type: integer, copy_to: _all }
+      compression_mode: { type: keyword, copy_to: _all }
+      width: { type: integer, copy_to: _all }
+      height: { type: integer, copy_to: _all }
+      pixel_aspect_ratio: { type: float, copy_to: _all }
+      display_aspect_ratio: { type: keyword, copy_to: _all }
+      rotation: { type: float, copy_to: _all }
+      frame_rate_mode: { type: keyword, copy_to: _all }
+      frame_rate: { type: float, copy_to: _all }
+      frame_count: { type: integer, copy_to: _all }
+      resolution: { type: integer, copy_to: _all }
+      colorimetry: { type: keyword, copy_to: _all }
+      color_space: { type: keyword, copy_to: _all }
+      chroma_subsampling: { type: keyword, copy_to: _all }
+      bit_depth: { type: integer, copy_to: _all }
+      scan_type: { type: keyword, copy_to: _all }
+      interlacement: { type: keyword, copy_to: _all }
+      bits_pixel_frame: { type: float, copy_to: _all }
+      stream_size: { type: long, copy_to: _all }
+      proportion_of_this_stream: { type: float, copy_to: _all }
+      header_size: { type: long, copy_to: _all }
+      data_size: { type: long, copy_to: _all }
+      footer_size: { type: long, copy_to: _all }
+      language: { type: keyword, copy_to: _all }
+      color_primaries: { type: keyword, copy_to: _all }
+      transfer_characteristics: { type: keyword, copy_to: _all }
+      matrix_coefficients: { type: keyword, copy_to: _all }
+      is_streamable: { type: boolean, copy_to: _all }
+      writing_application: { type: keyword, copy_to: _all }
+      file_last_modification_date: { type: date, copy_to: _all }
+      file_last_modification_date_local: { type: date, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -223,84 +223,84 @@ mapping:
-      date_ingested: { type: date, include_in_all: false }
-      size: { type: long, include_in_all: false }
-      mime_type: { type: keyword, include_in_all: false }
+      puid: { type: keyword, copy_to: _all }
+      filename: { type: keyword, copy_to: _all }
+      last_modified: { type: date, copy_to: _all }
+      date_ingested: { type: date, copy_to: _all }
+      size: { type: long, copy_to: _all }
+      mime_type: { type: keyword, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -630,5 +630,5 @@ mapping:
-      start_date_sort: { type: date, include_in_all: false }
-      end_date_sort: { type: date, include_in_all: false }
+      start_date_sort: { type: date, copy_to: _all }
+      end_date_sort: { type: date, copy_to: _all }
```
Same for a lot of fields, there is a lot. If the field had `include_in_all: false` it should not have `copy_to: _all`. And, looking at the condition, we only care about `text` types, so we only need to add `copy_to: _all` to those when needed.

jraddaoui reviewed (commented):
Thanks @anvit! This looks a lot clearer to me now.

anvit committed:
Change references to index type for ElasticSearch

Update any references to ElasticSearch index type to index name.
Also add getType as an alias for getIndex for backwards compatibility
with custom themes that still might be making references to
ElasticSearch's getType.

anvit committed:
Fix copy to _all fields for ES mapping

Removed copy_to _all from all fields that previously had include_in_all
set to false.

anvit committed:
Remove unnecessary ElasticSearch type variables

Remove unnecessary $type ElasticSearch variables in QubitLftSyncer,
arElasticSearchPlugin, and updatePublicationStatusTask

anvit committed:
Remove ES index prefix from MultiIndexWrapper

Remove the index prefix from arElasticSearchMultiIndexWrapper and use
AtoM class names internally instead of using prefix to refer to them.

jraddaoui commented on apps/qubit/modules/search/actions/autocompleteAction.class.php:
```diff
@@ -76,7 +76,12 @@ public function execute($request)
         foreach ($items as $item) {
             $search = new \Elastica\Search($client);
-            $search->addIndex($index)->addType($index->getType($item['type']));
+            foreach ($indices as $type => $index) {
+                $elasticSearchTypeName = QubitSearch::getInstance()::ES_TYPE;
+
+                // This can be updated in ES 7.x when type params are optional
+                $search->addIndex($index)->addType($index->getType($elasticSearchTypeName));
+            }
```
This works the same for me:

```php
foreach ($indices as $index) {
    $search->addIndex($index)->addType(QubitSearch::getInstance()::ES_TYPE);
}
```

And I think the type is already optional here, this also works for me:

```php
foreach ($indices as $index) {
    $search->addIndex($index);
}
```

PS. Somehow, actors are duplicated and shown as repositories in the results:

![image](https://github.com/user-attachments/assets/00b476fc-fc8b-47e4-b507-8fe22f3517db)

jraddaoui replied:
I found the issue, this is now adding all indices to all searches. We should get the wrapper before the loop, instead of getting all indices:

```php
$client = QubitSearch::getInstance()->client;
$indexWrapper = QubitSearch::getInstance()->index;
```

And only add the expected index to each search inside the loop:

```php
$search = new \Elastica\Search($client);
$search->addIndex($indexWrapper->getIndex($item['type']));
```

jraddaoui replied:
And a couple of extra details after that:

- Changing this and the for loops in the initialize function makes `getIndices()` unnecessary in the index wrapper class.
- This should be noted as something that could break custom themes extending the autocomplete.

jraddaoui commented on lib/QubitLftSyncer.class.php:
```diff
@@ -92,8 +92,8 @@ private function repairEsChildrenLftValues()
         $bulk = new Elastica\Bulk(QubitSearch::getInstance()->client);
-        $bulk->setIndex(QubitSearch::getInstance()->index->getName());
-        $bulk->setType('QubitInformationObject');
+        $bulk->setIndex(QubitSearch::getInstance()->index->getIndex('QubitInformationObject')->getName());
```
You could pass the index directly, without calling getName:

https://github.com/ruflin/Elastica/blob/6.x/lib/Elastica/Bulk.php#L50-L64

jraddaoui commented on lib/QubitLftSyncer.class.php:
```diff
@@ -92,8 +92,8 @@ private function repairEsChildrenLftValues()
-        $bulk->setIndex(QubitSearch::getInstance()->index->getName());
-        $bulk->setType('QubitInformationObject');
+        $bulk->setIndex(QubitSearch::getInstance()->index->getIndex('QubitInformationObject')->getName());
+        $bulk->setType(QubitSearch::getInstance()::ES_TYPE);
```
Looking at the usage of `_type` in https://github.com/ruflin/Elastica/blob/6.x/lib/Elastica/Bulk.php I bet we don't need to set it here either, but I have not tested it.

jraddaoui replied:
Ignore this, I just tested it and it looks like it's still needed.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,83 @@
+ * and has methods that match signatures of pre ES 6.x methods that used a
+ * single index with multiple types instead of multiple index with a single
+ * type or no type.
+ */
```
Maybe add a note that we have `indices` here in a separate class instead of moving it to arElasticSearchPlugin class to maintain backwards compatibility in client themes, similar to the getType function, but as a whole for the entire class.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -355,7 +385,7 @@ public function addDocument($data, $type)
             }
         } else {
-            $this->index->getType($type)->addDocument($document);
+            $this->index->getIndex($indexName)->addDocuments([$document]);
```
Why getting the index instead of using the wrapper directly like in flushBatch?

```php
$this->index->addDocuments($indexName, [$document]);
```

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -379,10 +409,16 @@ public function partialUpdate($object, $data)
+
         try {
-            $this->index->getType(get_class($object))->updateDocument($document);
+            $this->index->getIndex($indexName)->updateDocuments([$document]);
```
Going with the wrapper in the end, we may want to add methods for the update and the partial updates in there too. Not a big deal for sure.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,57 +541,84 @@ protected function initialize()
+                    // and will have be removed by 9.x when it is discontinued
+                    $mapping->send();
+                }
             }
```
There are two unnecessary loops that make this a lot harder to read. If you make the ES index a variable in the first loop, there is no need to loop over getIndices nor the mapping again:

```php
foreach ($this->mappings as $indexName => $indexProperties) {
    $indexName = 'Qubit'.sfInflector::camelize($indexName);
    $prefixedIndexName = $this->config['index']['name'].'_'.strtolower($indexName);
    $index = $this->client->getIndex($prefixedIndexName);
    $this->index->addIndex($indexName, $index);

    try {
        $index->open();
    } catch (Exception $e) {
        // If the index has not been initialized, create it
        if ($e instanceof \Elastica\Exception\ResponseException) {
            // Removed for the review
        }

        // Define mapping in elasticsearch
        $mapping = new \Elastica\Type\Mapping();

        // Setting a dummy type since it is required in ES 6.x
        // but it can be removed in 7.x when it becomes optional
        $mapping->setType($index->getType(self::ES_TYPE));
        $mapping->setProperties($indexProperties['properties']);

        // Parse other parameters
        unset($this->mapping[$indexName]->indexProperties['properties']);
        foreach ($this->mapping[$indexName]->indexProperties as $key => $value) {
            $mapping->setParam($key, $value);
        }

        $this->log(sprintf('Defining mapping %s...', $indexName));

        // In ES 7.x this should be changed to:
        // $mapping->send($index, [ 'include_type_name' => false ])
        // which can be removed in 8.x since that is the default behaviour
        // and will have be removed by 9.x when it is discontinued
        $mapping->send();
    }
}
```

jraddaoui replied:
Also:

- It feels rare that we define the mapping outside the `if`.
- We may want to use the prefixed index name in the log:

  ```php
  $this->log(sprintf('Defining mapping for index %s...', $prefixedIndexName));
  ```

anvit replied:
@jraddaoui it does feel odd to define the mappings outside the `if`. I let it be because I wasn't sure if there was some historical reason or if there was a scenario where the index existed but could not be opened because it didn't have mappings declared. In my testing changing it to:
```
try {                                                             
  $index->open();                                      
} catch (\Elastica\Exception\ResponseException $e) {
// If the index has not been initialized, create it
```
didn't seem to cause any problems (both in the scenario when the index existed and when it didn't). Let me know if you want me to make this change

jraddaoui replied:
Yeah, changing the catch and removing the if makes sense to me.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -605,9 +680,9 @@ private function displayTypesToIndex($excludeTypes)
         $this->log('Types that will be indexed:');
 
-        foreach ($this->mappings as $typeName => $typeProperties) {
-            if (!in_array(strtolower($typeName), $excludeTypes)) {
-                $this->log(' - '.$typeName);
+        foreach ($this->mappings as $indexName => $indexProperties) {
+            if (!in_array(strtolower($indexName), $excludeTypes)) {
+                $this->log(' - '.$indexName);
                 ++$typeCount;
             }
         }
```
We may want to change this too, "Indices that will be updated." and use the prefixed index name.

jraddaoui replied:
Okay, maybe not here. These need to match the `exclude-types` argument from the search:populate task.

Now that I think about that, we should probably revisit how is that done with multiple indexes instead of types.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -32,11 +32,11 @@ mapping:
-      location: { type: geo_point }
+      postal_code: { type: text }
+      country_code: { type: keyword }
+      location: { type: geo_point, copy_to: _all }
```
The condition checks the type too, making sure it's a `text` type field. It doesn't matter if you add it here, but it won't be copied to _all in the end and adds some confusion.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -61,22 +61,22 @@ mapping:
+      start_date: { type: date }
+      end_date: { type: date }
+      start_date_string: { type: keyword, copy_to: _all }
+      end_date_string: { type: keyword, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -91,8 +91,8 @@ mapping:
-      date: { type: date, include_in_all: false }
-      date_string: { type: keyword }
+      date: { type: date }
+      date_string: { type: keyword, copy_to: _all }
```
Same.

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -102,111 +102,111 @@ mapping:
     dynamic: strict
     properties:
-      slug: { type: keyword }
+      slug: { type: keyword, copy_to: _all }
```
I think you get the point ;)

jraddaoui reviewed (commented):
Thanks @anvit! Getting there, just a few minor tweaks!

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -427,12 +463,24 @@ public function delete($object)
         }
 
         if ($this->batchMode) {
+            $indexName = get_class($object);
```
I just noticed, this needs to happen before the `if`, as it's also used in the `else`.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,83 @@
+        $this->indices[$name] = $index;
+    }
+
+    public function delete()
+    {
+        foreach ($this->indices as $index) {
+            $index->delete();
+        }
+    }
```
Given all the feedback above, this would be the only function without a `$name` parameter in the wrapper (apart from the constructor). And, when we revisit the `exclude-types` option I think we'll add that parameter.

It makes a lot of sense to me that all methods from this class require a name parameter now that it's a multi index wrapper instead of an index decorator.

It would be even clearer if `$index` was called `$indicesWrapper` in `arElasticSearchPlugin`, but preventing issues in custom code is a stronger point than clarity.

Just a note.

anvit committed:
Fix autocompleteAction bug, address CR feedback

Fix bug in search autocompleteAction which was showing duplicate results
for actors under repositories. Also address some of the CR feedback and
update/add methods for add, update, and deleteById in the
arElasticSearchMultiIndexWrapper.

anvit committed:
Remove non text fields from _all ES mapping

anvit committed:
Remove unnecessary loop in arElasticSearchPlugin

Remove unnecessary foreach loop in arElasticSearchPlugin. Also move code
for adding filter for stripping md tags into a seperate method to make
the initialization code more readable.

jraddaoui commented on apps/qubit/modules/search/actions/autocompleteAction.class.php:
```diff
@@ -76,7 +75,8 @@ public function execute($request)
         foreach ($items as $item) {
             $search = new \Elastica\Search($client);
-            $search->addIndex($index)->addType($index->getType($item['type']));
+            $indexWrapper = QubitSearch::getInstance()->index;
```
Better if we get the wrapper only once before the loop.

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,61 +540,84 @@ protected function initialize()
                 // Parse other parameters
-                unset($typeProperties['properties']);
-                foreach ($typeProperties as $key => $value) {
+                unset($this->mapping[$indexName]->indexProperties['properties']);
+                foreach ($this->mapping[$indexName]->indexProperties as $key => $value) {
```
Can't you do this directly over the `$indexProperties` variable? Why accessing the mapping again?

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchPlugin.class.php:
```diff
@@ -493,61 +540,84 @@ protected function initialize()
+                $this->config['index']['configuration']['analysis']['analyzer'][$key]['char_filter'] = $filters;
+            }
+        }
+    }
```
Nice move!

jraddaoui commented on plugins/arElasticSearchPlugin/config/mapping.yml:
```diff
@@ -630,5 +630,5 @@ mapping:
-      start_date_sort: { type: date, include_in_all: false }
-      end_date_sort: { type: date, include_in_all: false }
+      start_date_sort: { type: date }
+      end_date_sort: { type: date }
```
Neat! A lot clearer compared to the previous version.

jraddaoui reviewed (commented):
Nice, thanks @anvit!

jraddaoui commented on plugins/arElasticSearchPlugin/lib/arElasticSearchMultiIndexWrapper.class.php:
```diff
@@ -0,0 +1,96 @@
+    public function getIndices()
+    {
+        return $this->indices;
+    }
```
This function is no longer used.

anvit committed:
Update ES index refresh to only refresh one index

Update the index refresh method to accept an optional name param that
only refreshes that particular index in
arElasticSearchMultiIndexWrapper. Also address some small CR feedback.

anvit committed:
Refactor ES initialize and populate methods (WIP)

jraddaoui committed:
Tweak index recreation

jraddaoui committed:
Initialize on populate

jraddaoui committed:
Tweak flush/refresh operation

jraddaoui committed:
Only ignore 404 errors when deleting indexes

jraddaoui committed:
Prevent flush if current index name is not set

jraddaoui committed:
Make loadDiacriticsMappings a private function

jraddaoui committed:
Improve populate logging

jraddaoui committed:
Configure filters once before indices creation

jraddaoui committed:
Normalize loadDiacriticsMappings function

Check setting and update config global variable there.

jraddaoui committed:
Create indices on update if they don't exist

jraddaoui committed:
Trust recreate option on index creation

jraddaoui committed:
Remove unused function
